File: linked_list/singly_test.py
import unittest
from singly import Node, SinglyLinkedList

class TestSinglyLinkedList(unittest.TestCase):
  def setUp(self):
    self.list = SinglyLinkedList()
    self.list.insert(1)
    self.list.insert(2)
    self.list.insert(3)
    self.list.insert(4)
    self.list.insert(5)
    self.list.append(6)

  def test_insert(self):
    self.list.insert(100)
    self.assertEqual(self.list.head.data, 100)
    self.assertNotEqual(self.list.head.data, 5)

  def test_includes(self):
    self.assertEqual(self.list.includes(3), True)
    self.assertEqual(self.list.includes(5), True)
    self.assertEqual(self.list.includes(6), True)
    
    self.assertEqual(self.list.includes(100), False)
    self.assertEqual(self.list.includes('hello'), False)

  def test_delete(self):
    self.assertEqual(self.list.includes(3), True)
    self.list.delete(3)
    self.assertEqual(self.list.includes(3), False)

    # Deleting the head
    self.list.delete(5)
    self.assertNotEqual(self.list.head.data, 5)
    self.assertEqual(self.list.head.data, 4)

    x = self.list.delete(10000)
    self.assertEqual(x, "not found")
  
if __name__ == '__main__':
    unittest.main()

File: linked_list/singly_test.go
package linked_list

import (
	"testing"
)

func Setup() LinkedList {
	node3 := Node{
		Data:  25,
		Child: nil,
	}
	node2 := Node{
		Data:  17,
		Child: &node3,
	}
	node1 := Node{
		Data:  3,
		Child: &node2,
	}
	head := Node{
		Data:  5,
		Child: &node1,
	}
	list := LinkedList{Head: &head}
	return list
}

func TestInsert(t *testing.T) {
	list := Setup()
	length := list.Length()
	tests := []int{10, 11, 12, 13, 14, 100, 101}

	for _, test := range tests {
		list.Insert(test)
		if list.Head.Data != test {
			t.Errorf("Insert test failed")
		}
		if list.Length() != length+1 {
			t.Errorf("Insert test failed")
		}
		length++
	}
}

func TestLength(t *testing.T) {
	list := Setup()
	if list.Length() != 4 {
		t.Errorf("Length function failed")
	}
}

func TestSearch(t *testing.T) {
	list := Setup()
	if !list.Search(list.Head, 17) {
		t.Errorf("Search function failed")
	}
	if list.Search(list.Head, 1000) {
		t.Errorf("Search function failed")
	}
}

func TestAppend(t *testing.T) {
	list := Setup()
	newNode := Node{
		Data:  23,
		Child: nil,
	}
	newNode2 := Node{
		Data:  99,
		Child: nil,
	}

	list.Append(list.Head, &newNode)
	list.Append(list.Head, &newNode2)
	if list.Length() != 6 || !list.Search(list.Head, 99) || !list.Search(list.Head, 23) {
		t.Errorf("Append failed")
	}
}

func TestAppendEmptyList(t *testing.T) {
	newHead := Node{
		Data:  0,
		Child: nil,
	}
	newList := LinkedList{Head: &newHead}

	node99 := Node{
		Data:  99,
		Child: nil,
	}
	node98 := Node{
		Data:  98,
		Child: nil,
	}
	newList.Append(newList.Head, &node98)
	newList.Append(newList.Head, &node99)
	if newList.Length() != 3 || !newList.Search(newList.Head, 99) || !newList.Search(newList.Head, 98) {
		t.Errorf("Append failed")
	}
}

func TestDelete(t *testing.T) {
	list := Setup()
	list.Delete(list.Head, 17)
	if list.Length() != 3 || list.Search(list.Head, 17) {
		t.Errorf("Delete failed")
	}
}

func TestContents(t *testing.T) {
	list := Setup()
	if !list.Contents() {
		t.Errorf("Contents failed")
	}
}
